src/common/io/tls/client.c

```diff
@@ -0,0 +1,550 @@
+***********************************************************************************************************************************/
+size_t
+tlsClientRead(TlsClient *this, Buffer *buffer)
+{
+    FUNCTION_DEBUG_BEGIN(logLevelTrace);
+        FUNCTION_DEBUG_PARAM(TLS_CLIENT, this);
+        FUNCTION_DEBUG_PARAM(BUFFER, buffer);
+
+        FUNCTION_TEST_ASSERT(this != NULL);
+        FUNCTION_TEST_ASSERT(this->session != NULL);
+        FUNCTION_TEST_ASSERT(buffer != NULL);
+        FUNCTION_TEST_ASSERT(!bufFull(buffer));
+    FUNCTION_DEBUG_END();
+
+    ssize_t actualBytes = 0;
+
+    // If no tls data pending then check the socket
+    if (!SSL_pending(this->session))
+    {
+        // Initialize the file descriptor set used for select
+        fd_set selectSet;
+        FD_ZERO(&selectSet);
+        // We know socket is not negative because it passed error handling, so it is safe to cast to unsigned
+        FD_SET((unsigned int)this->socket, &selectSet);
+
+        // Initialize timeout struct used for select.  Recreate this structure each time since Linux at least will modify it.
+        struct timeval timeoutSelect;
+        timeoutSelect.tv_sec = (time_t)(this->timeout / MSEC_PER_SEC);
+        timeoutSelect.tv_usec = (time_t)(this->timeout % MSEC_PER_SEC * 1000);
+
+        // Determine if there is data to be read
+        int result = select(this->socket + 1, &selectSet, NULL, NULL, &timeoutSelect);
+        THROW_ON_SYS_ERROR_FMT(result == -1, AssertError, "unable to select from '%s:%u'", strPtr(this->host), this->port);
+
+        // If no data read after time allotted then error
+        if (!result)
+        {
+            THROW_FMT(
+                FileReadError, "unable to read data from '%s:%u' after %" PRIu64 "ms",
+                strPtr(this->host), this->port, this->timeout);
+        }
+    }
+
+    // Read and handle errors
+    size_t expectedBytes = bufRemains(buffer);
+    actualBytes = SSL_read(this->session, bufRemainsPtr(buffer), (int)expectedBytes);
+
+    cryptoError(actualBytes < 0, "unable to read from TLS");
+
+    // Update amount of buffer used
+    bufUsedInc(buffer, (size_t)actualBytes);
+
+    // If zero bytes were returned then the connection was closed
+    if (actualBytes == 0)
+        tlsClientClose(this);
+
+    FUNCTION_DEBUG_RESULT(SIZE, (size_t)actualBytes);
+}
+
+/***********************************************************************************************************************************
+Write to the tls session
+***********************************************************************************************************************************/
+void
+tlsClientWrite(TlsClient *this, const Buffer *buffer)
+{
+    FUNCTION_DEBUG_BEGIN(logLevelTrace);
+        FUNCTION_DEBUG_PARAM(TLS_CLIENT, this);
+        FUNCTION_DEBUG_PARAM(BUFFER, buffer);
+
+        FUNCTION_TEST_ASSERT(this != NULL);
+        FUNCTION_TEST_ASSERT(this->session != NULL);
+    FUNCTION_DEBUG_END();
```

Comment:
Shouldn't buffer be required and not NULL, so is there a missing ASSERT here?

Reply:
Yes, fixed.